Automattic/wp-calypso
# Use checkstyle to report tsc errors

## Background

We have some TypeScript files inside `./client/`, but we are not typechecking them (except for those inside `./client/landing/gutenboarding`).

When we typecheck them, there are a ton of errors.

## Changes

The main goal is to have a [ratcheted build](https://robertgreiner.com/continuous-code-improvement-using-ratcheting/) that ensures the problem doesn't get worse while we work on improving it. 

The main change is use a `tsc` -> `checkstyle` conversion tool, and let TeamCity parse it and keep track of the number of errors. Then we configure the build to fail in the number of errors increase.

Other minor changes are:

- Delete `./client/landing/gutenboarding/tsconfig.json` so it doesn't have a special status anymore, it is just part of the `./client` TS project.

## Testing instructions

Check the "Unit tests" build for this branch. It should:
- Be green
- Have a 'Code inspection' tab with tsc errors
- Have a `typescript_checkstyle.xml` artifact

## Update

I've left the config to fail the build if the number of errors increases. I think that should be the end goal, but probably it is better to have the build gathering info without failing it for a while to make sure it is stable.
